File: binary_tree_DFS/longest_zigzag_path/longest_zigzag_path.py
def longestZigZag(root):
    longestSteps = 0
    def dfs(node, goLeft, steps):
        if node:
            nonlocal longestSteps
            logestSteps = max(steps, logestSteps)
            if goLeft:
                dfs(node.left, False, steps + 1)
                dfs(node.right, True, 1)
            else:
                dfs(node.right, True, steps + 1)
                dfs(node.left, False, 1)
        else:
            return
        
    dfs(root, True, 0)
    dfs(root, False, 0)
    return longestSteps

File: binary_tree_DFS/longest_zigzag_path/longest_zigzag_path.js
/*
start from any node, end at any node
traverse function
helper function to calculate longest path
    - if node === null, return 0
    - if node exist, longestPath++
    - declare goLeft = true
    - if node.left === null, goLeft = false
*/

const longestZigZag = (root) => {
    let longestPath = 0
    
    const dfs = (node, goLeft, steps) => {
        if (node) {
            longestPath = Math.max(steps, longestPath)

            if (goLeft) {
                dfs(node.left, false, steps + 1)
                dfs(node.right, true, 1)
            } else {
                dfs(node.right, true, steps + 1)
                dfs(node.left, false, 1)
            }
        }
    }

    dfs(root, true, 0)
    dfs(root, false, 0)
    
    return longestPath
}